### PR TITLE
Feature/data serviceendpoint azurerm

### PR DESCRIPTION
**Pull request by a30004053**
## All Submissions:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] I have updated the documentation accordingly.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.
* [x] My code follows the code style of this project.
* [x] I ran lint checks locally prior to submission.
* [x] Have you checked to ensure there aren't other open PRs for the same update/change?

## What about the current behavior has changed?

This PR adds a new data source `azuredevops_serviceendpoint_azurerm` which returns details of an AzureRM service endpoint defined by an `id` or a `service_endpoint_name`

## Does this introduce a change to `go.mod`, `go.sum` or `vendor/`?

- [ ] Yes
- [x] No

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

## Any relevant logs, error output, etc?

https://pst.klgrth.io/paste/wesap

## Other information

Acceptance Test: `TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource`
```bash
=== RUN   TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
=== PAUSE TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
=== CONT  TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
--- PASS: TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource (51.29s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        51.399s
```
Acceptance Test: `TestAccServiceEndpointAzureRM_with_serviceEndpointName_DataSource`
```bash
=== RUN   TestAccServiceEndpointAzureRM_with_serviceEndpointName_DataSource
=== PAUSE TestAccServiceEndpointAzureRM_with_serviceEndpointName_DataSource
=== CONT  TestAccServiceEndpointAzureRM_with_serviceEndpointName_DataSource
--- PASS: TestAccServiceEndpointAzureRM_with_serviceEndpointName_DataSource (51.61s)
PASS
ok      github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/acceptancetests        51.719s
```

### CODE REVIEW

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+		)
+		if err != nil {
+			if utils.ResponseWasNotFound(err) {
+				d.SetId("")
+				return nil
+			}
+			return fmt.Errorf("Error looking up service endpoint with ID (%v) and projectID (%v): %v", serviceEndpointID, projectID, err)
+		}
+
+		if serviceEndpoint.Id == nil {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: `serviceEndpoint.Id`  cannot be `nil` unless there is service bug, this if statement can removed.

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+			d.SetId("")
+		} else {
+			// prevent credential leak into data source
+			(*serviceEndpoint.Data)["creationMode"] = ""
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: service will not return any of the sensitive data, all of the them will be empty string in the response body.

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+				Type:         schema.TypeString,
+				Optional:     true,
+				Computed:     true,
+				ExactlyOneOf: []string{"service_endpoint_name", "id"},
+				ValidateFunc: validation.IsUUID,
+			},
+
+			"authorization": {
+				Type:         schema.TypeMap,
+				Optional:     true,
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: `authorization` is not required to get the service  connection. `optional=true` can be removed.

**Review by xuzhang3 — CHANGES_REQUESTED** [review]

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go (hunk 2 above)
Comment (a30004053) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: keeping `(*serviceEndpoint.Data)["creationMode"] = ""` as the `creationMode` key is required in order to use the `flattenServiceEndpointAzureRM` func.

The rest of your comment has been addressed

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+func DataServiceEndpointAzureRM() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceServiceEndpointAzureRMRead,
+		Schema: map[string]*schema.Schema{
+			"project_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"service_endpoint_name": {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go:24]: suggest rename `service_endpoint_name` to `name`

Comment (a30004053) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go:24]: renamed the attribute, also added `service_endpoint_name` as a computed value, failing which the `doBaseFlattening` func panics during Acceptance testing
```bash
=== RUN   TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
=== PAUSE TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
=== CONT  TestAccServiceEndpointAzureRM_with_serviceEndpointID_DataSource
panic: Invalid address to set: []string{"service_endpoint_name"}

goroutine 254 [running]:
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*ResourceData).Set(0xc0001b5700, {0xf4aa3b, 0x15}, {0xd864a0, 0xc0006fe540})
        /root/terraform-provider-azuredevops/vendor/github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema/resource_data.go:230 +0x291
github.com/microsoft/terraform-provider-azuredevops/azuredevops/internal/service/serviceendpoint.doBaseFlattening(0xda8080?, 0xc000786bd0, 0xc00082cb60)
```

File: azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+func DataServiceEndpointAzureRM() *schema.Resource {
+	return &schema.Resource{
+		Read: dataSourceServiceEndpointAzureRMRead,
+		Schema: map[string]*schema.Schema{
+			"project_id": {
+				Type:     schema.TypeString,
+				Required: true,
+			},
+
+			"name": {
```
Comment (xuzhang3) [inline azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: `name` and `service_endpoint_name` is duplicate, can we add a new flatten function for this data source.   What I thinking is the resource name has the `service endpoint` included, we can simplify the `service_endpoint_name` to `name`.

Comment (a30004053) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: Agreed. Considered doing that while writing this but wanted to keep the code as DRY as possible.

Comment (xuzhang3) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: but the `name` and `service_endpoint_name`  do break DRY somehow,  we can remove `name ` and keep only `service_endpoint_name`

Comment (a30004053) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: We break DRY one way or the other :), personally, I think we can continue with `service_endpoint_name` like the original datasource, thoughts?

Comment (xuzhang3) [reply azuredevops/internal/service/serviceendpoint/data_serviceendpoint_azurerm.go]: agreed,  continue with `service_endpoint_name `

**Comment by xuzhang3** [comment]
LGTM

**Pull request merged**